### PR TITLE
engine post

### PR DESCRIPTION
**Pull request by tobymao**

### CODE REVIEW

File: posts/python_sql_engine.md [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+  FROM y AS y
+  WHERE TRUE
+  GROUP BY y.a
+) AS "_u_0"
+  ON x.a = "_u_0".a
+WHERE ("_u_0".a = 1 AND NOT "_u_0".a IS NULL)'
+```
+
+### pushdown_predicates
+Push down filters into the innermost query.
```
Comment (GeorgeSittas) [inline posts/python_sql_engine.md:131]:
```suggestion
Pushes down filters into the innermost query.
```

File: posts/python_sql_engine.md [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+- Ensure each column is unambiguous and expand stars.
+
+```sql
+SELECT * FROM x`;
+
+SELECT "db"."x" AS "x";
+```
+
+### simplify
+Boolean and math simplification. Check out all the [test cases](https://github.com/tobymao/sqlglot/blob/main/tests/fixtures/optimizer/simplify.sql).
```
Comment (GeorgeSittas) [inline posts/python_sql_engine.md:84]:
```suggestion
Simplifies math expressions. Check out all the [test cases](https://github.com/tobymao/sqlglot/blob/main/tests/fixtures/optimizer/simplify.sql).
```

File: posts/python_sql_engine.md [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+SELECT *
+FROM (
+  SELECT *
+  FROM x AS x
+  WHERE y.a = 1
+) AS y WHERE TRUE
+```
+
+### annotate_types
+Infer all types throughout the AST given schema information and function type definitions.
```
Comment (GeorgeSittas) [inline posts/python_sql_engine.md:149]:
```suggestion
Infers all types throughout the AST given schema information and function type definitions.
```

File: posts/python_sql_engine.md [hunk 4]
```diff
@@ -0,0 +1,187 @@
+# Writing a Python SQL Engine from Scratch
+[Toby Mao](https://www.linkedin.com/in/toby-mao/)
+
+## Introduction
+
+When I first started writing SQLGlot in early 2021, my goal was just to translate SQL queries from SparkSQL to Presto and vice versa. However, over the last year and a half due to scope creep, I've ended up with a full fledge SQL engine. SQLGlot can now parse and transpile between [18 SQL dialects](https://github.com/tobymao/sqlglot/blob/main/sqlglot/dialects/__init__.py) and can execute all 24 [TPC-H](https://www.tpc.org/tpch/) SQL queries. The parser and engine are all handwritten from scratch with Python.
```
Comment (GeorgeSittas) [inline posts/python_sql_engine.md:6]:
```suggestion
When I first started writing SQLGlot in early 2021, my goal was just to translate SQL queries from SparkSQL to Presto and vice versa. However, over the last year and a half due to scope creep, I've ended up with a full fledged SQL engine. SQLGlot can now parse and transpile between [18 SQL dialects](https://github.com/tobymao/sqlglot/blob/main/sqlglot/dialects/__init__.py) and can execute all 24 [TPC-H](https://www.tpc.org/tpch/) SQL queries. The parser and engine are all handwritten from scratch with Python.
```

File: posts/python_sql_engine.md [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+I started working on SQLGlot because of my work on the [experimentation and metrics platform](https://netflixtechblog.com/reimagining-experimentation-analysis-at-netflix-71356393af21) at Netflix, where I built tools that allowed data scientists to define and compute SQL based metrics. Netflix relied on multiple engines to query data (Spark, Presto, and Druid) so my team built the metrics platform around [PyPika](https://github.com/kayak/pypika), a Python SQL query builder so that definitions could be reused across multiple engines. However, it became quickly apparent that writing python code to programatically generate SQL was challenging for data scientists, especially those with academic backgrounds since they were mostly familiar with R and SQL. But at the time, the only Python SQL parser was [sqlparse]([https://github.com/andialbrecht/sqlparse) which is not actually a parser, but a tokenizer, so having users write raw SQL into the platform wasn't really an option. Some time later, I randomly stumbled across [Crafting Interpreters](https://craftinginterpreters.com/) and realized that I could use it as a guide towards creating my on SQL parser / transpiler.
```
Comment (GeorgeSittas) [inline posts/python_sql_engine.md:8]:
```suggestion
I started working on SQLGlot because of my work on the [experimentation and metrics platform](https://netflixtechblog.com/reimagining-experimentation-analysis-at-netflix-71356393af21) at Netflix, where I built tools that allowed data scientists to define and compute SQL based metrics. Netflix relied on multiple engines to query data (Spark, Presto, and Druid) so my team built the metrics platform around [PyPika](https://github.com/kayak/pypika), a Python SQL query builder, which enabled the reuse of definitions across multiple engines. However, it became quickly apparent that writing python code to programatically generate SQL was challenging for data scientists, especially those with academic backgrounds since they were mostly familiar with R and SQL. But at the time, the only Python SQL parser was [sqlparse]([https://github.com/andialbrecht/sqlparse) which is not actually a parser, but a tokenizer, so having users write raw SQL into the platform wasn't really an option. Some time later, I randomly stumbled across [Crafting Interpreters](https://craftinginterpreters.com/) and realized that I could use it as a guide towards creating my own SQL parser / transpiler.
```

File: posts/python_sql_engine.md [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+
+## Tokenizing
+
+The first step is to convert the sql string into a list of tokens. SQLGlot's tokenizer is quite simple and can be found [here](https://github.com/tobymao/sqlglot/blob/main/sqlglot/tokens.py#L763). In a while loop, it checks each character and either appends the character to the current token or makes a new token.
+
+Running the SQLGlot tokenizer shows the output.
+
+![Tokenizer Output](python_sql_engine_images/tokenizer.png)
+
+Each keyword has been converted to a SQLGlot Token object. Each token has some metadata associated with it like line / column information for error messages. Comments are also a part of the token so that comments can be preserved.
```
Comment (GeorgeSittas) [inline posts/python_sql_engine.md:45]:
```suggestion
Each keyword has been converted to a SQLGlot Token object. Each token has some metadata associated with it like line / column information for error messages. Comments are also a part of the token so that they can be preserved when transpiling SQL.
```

File: posts/python_sql_engine.md [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+
+Similar to the tokenizer, it consumes the tokens sequentially but it uses a recursive algorithm instead. The tokens are converted into a single AST node which presents the SQL query. The SQLGlot parser was designed to support various dialects so it contains many options for overriding parsing functionality.
+
+![Parser Output](python_sql_engine_images/parser.png)
+
+The AST is a generic representation of a given SQL query. Each dialect can override or implement its own generator that can convert an AST object into syntatically correct SQL.
+
+## Optimizing
+
+Once we have our AST, we need to simplify it and put it into a consistent form to actually run queries efficiently. For example, how do you know which table column 'b' comes from in `SELECT bar.a, b FROM bar JOIN baz`. Given a schema, we can infer that b comes from table baz. When optimizing queries, most engines first convert the AST into a logical plan and then optimize the plan. However, I chose to **optimize the AST directly** for two reasons.
```
Comment (GeorgeSittas) [inline posts/python_sql_engine.md:59]:
```suggestion
Once we have our AST, we need to simplify it and transform it into a canonical form that enables us to run queries efficiently. For example, how do you know which table column 'b' comes from in `SELECT bar.a, b FROM bar JOIN baz`. Given a schema, we can infer that b comes from table baz. When optimizing queries, most engines first convert the AST into a logical plan and then optimize the plan. However, I chose to **optimize the AST directly** for two reasons.
```

File: posts/python_sql_engine.md [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+
+![Optimizer Output](python_sql_engine_images/optimizer.png)
+
+The optimizer currently has [17 rules](https://github.com/tobymao/sqlglot/tree/main/sqlglot/optimizer). Each of these rules is applied, transforming the AST in place. The combination of these rules creates "canonical" sql which can then be more easily converted into a logical plan and then executed.
+
+Some example rules are:
+
+### qualify\_tables and qualify_columns
+- Adds all db/catalog qualifiers to tables and forces an alias.
+- Ensure each column is unambiguous and expand stars.
```
Comment (GeorgeSittas) [inline posts/python_sql_engine.md:75]:
```suggestion
- Adds all db/catalog qualifiers to tables and forces aliases.
- Ensures each column is unambiguous and expands star projections.
```

**Review by GeorgeSittas — APPROVED** [review]
I left a few suggestions here and there, but overall I think it's a great post!

One thing that stood out to me was that the images look big compared to the text. Is this the case for other people too? It may be more user-friendly to zoom them out a bit so that they can be viewed in one go instead of having to scroll, though that's probably affected by screen size now that I think about it.

**Review by eakmanrq — APPROVED** [review]
Great post! I could see how the images might seem big but it didn't stand out to me when I was reading it.

File: posts/python_sql_engine.md (hunk 4 above)
Comment (barakalon) [inline posts/python_sql_engine.md:6]:
```suggestion
When I first started writing SQLGlot in early 2021, my goal was just to translate SQL queries from SparkSQL to Presto and vice versa. However, over the last year and a half, I've ended up with a full-fledged SQL engine. SQLGlot can now parse and transpile between [18 SQL dialects](https://github.com/tobymao/sqlglot/blob/main/sqlglot/dialects/__init__.py) and can execute all 24 [TPC-H](https://www.tpc.org/tpch/) SQL queries. The parser and engine are all handwritten from scratch with Python.
```

Removed "due to scope creep" - I don't think it was necessary.

File: posts/python_sql_engine.md (hunk 5 above)
Comment (barakalon) [inline posts/python_sql_engine.md:8]: I feel like this paragraph can go under "Why?", leaving the intro a little more concise.

File: posts/python_sql_engine.md [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+
+Parts of SQLGlot's toolkit are being used today by [Ibis](https://github.com/ibis-project/ibis), [Quokka](https://github.com/marsupialtail/quokka), and [Splink](https://github.com/moj-analytical-services/splink).
```
Comment (barakalon) [inline posts/python_sql_engine.md:19]: Might not be as exciting of a use case, but https://github.com/kelsin/mysql-mimic uses the executor as-is.

File: posts/python_sql_engine.md [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+JOIN baz
+  ON bar.a = baz.a
+WHERE bar.a > 1
+```
+
+In this post, I'll walk through all the steps SQLGlot takes to run this query over Python objects.
+
+## Tokenizing
+
+The first step is to convert the sql string into a list of tokens. SQLGlot's tokenizer is quite simple and can be found [here](https://github.com/tobymao/sqlglot/blob/main/sqlglot/tokens.py#L763). In a while loop, it checks each character and either appends the character to the current token or makes a new token.
```
Comment (barakalon) [inline posts/python_sql_engine.md:39]: The line where the Tokenizer starts is likely to change throughout the life of this post. We can link to a specific commit or remove the line number.

File: posts/python_sql_engine.md (hunk 6 above)
Comment (barakalon) [reply posts/python_sql_engine.md:45]: The sentence about comments doesn't feel necessary.

File: posts/python_sql_engine.md (hunk 7 above)
Comment (barakalon) [inline posts/python_sql_engine.md:59]: While qualify_columns is super important to the optimizer, it's not really an optimization. I've seen this confuse people. I think we can just remove those two sentences. The paragraph will also flow better without it.

@GeorgeSittas - I think your change switches to passive voice. How about...

"Once we have our AST, we can transform it into an equivalent query that would produce the same results but execute for efficiently."

File: posts/python_sql_engine.md [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+
+![Parser Output](python_sql_engine_images/parser.png)
+
+The AST is a generic representation of a given SQL query. Each dialect can override or implement its own generator that can convert an AST object into syntatically correct SQL.
+
+## Optimizing
+
+Once we have our AST, we need to simplify it and put it into a consistent form to actually run queries efficiently. For example, how do you know which table column 'b' comes from in `SELECT bar.a, b FROM bar JOIN baz`. Given a schema, we can infer that b comes from table baz. When optimizing queries, most engines first convert the AST into a logical plan and then optimize the plan. However, I chose to **optimize the AST directly** for two reasons.
+
+1. I wanted a way to generate 'canonical sql'. Given two SQL statements, I wanted to be able to identify if they are semantically equivalent. For example `select 2` and `select 1 + 1` are functionally the same even though the SQL query is the same. Having the optimizer output SQL would allow me to convert 1 + 1 into 2.
```
Comment (barakalon) [inline posts/python_sql_engine.md:61]: I don't think this really captures the _why_, which might confuse people.

I think the why has to do with versioning Tobiko models. But that's hard to sum up quickly...

Perhaps we should move this after "It's easier to debug", and say something about "canonical sql" being super useful in other use cases, separate from the executor. E.g. determining if two queries are semantically equivalent.

File: posts/python_sql_engine.md [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+Some example rules are:
+
+### qualify\_tables and qualify_columns
+- Adds all db/catalog qualifiers to tables and forces an alias.
+- Ensure each column is unambiguous and expand stars.
+
+```sql
+SELECT * FROM x`;
+
+SELECT "db"."x" AS "x";
```
Comment (barakalon) [inline posts/python_sql_engine.md:80]:
```suggestion
SELECT x.a AS a, x.b AS b FROM db.x AS x;
```

File: posts/python_sql_engine.md [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+The optimizer currently has [17 rules](https://github.com/tobymao/sqlglot/tree/main/sqlglot/optimizer). Each of these rules is applied, transforming the AST in place. The combination of these rules creates "canonical" sql which can then be more easily converted into a logical plan and then executed.
+
+Some example rules are:
+
+### qualify\_tables and qualify_columns
+- Adds all db/catalog qualifiers to tables and forces an alias.
+- Ensure each column is unambiguous and expand stars.
+
+```sql
+SELECT * FROM x`;
```
Comment (barakalon) [inline posts/python_sql_engine.md:78]:
```suggestion
SELECT * FROM x;
```

File: posts/python_sql_engine.md [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+```
+
+### annotate_types
+Infer all types throughout the AST given schema information and function type definitions.
+
+## Planning
+After the SQL AST has been "optimized", it's much easier to [convert into a logical plan](https://github.com/tobymao/sqlglot/blob/main/sqlglot/planner.py). The AST is traversed and converted into a [DAG](https://en.wikipedia.org/wiki/Directed_acyclic_graph) consisting of one of five steps. The different types of steps are:
+
+### Scan
+The scan step selects columns from a table, applies projections, and finally filters the table.
```
Comment (barakalon) [inline posts/python_sql_engine.md:155]: I think we can just remove "The _ step" from all these

File: posts/python_sql_engine.md [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+
+The optimizer currently has [17 rules](https://github.com/tobymao/sqlglot/tree/main/sqlglot/optimizer). Each of these rules is applied, transforming the AST in place. The combination of these rules creates "canonical" sql which can then be more easily converted into a logical plan and then executed.
+
+Some example rules are:
+
+### qualify\_tables and qualify_columns
+- Adds all db/catalog qualifiers to tables and forces an alias.
+- Ensure each column is unambiguous and expand stars.
+
+```sql
```
Comment (barakalon) [inline posts/python_sql_engine.md:77]: We have a mix of inline code blocks and images.

Where do you plan to post this? Will it support inline code blocks with syntax highlighting?

File: posts/python_sql_engine.md [hunk 16]
```diff
@@ -0,0 +1,187 @@
+# Writing a Python SQL Engine from Scratch
+[Toby Mao](https://www.linkedin.com/in/toby-mao/)
+
+## Introduction
```
Comment (barakalon) [inline posts/python_sql_engine.md:4]: I think this is missing a summary of this post itself. That help readers know what to expect.

**Review by barakalon — COMMENTED** [review]
👍 I dig it

File: posts/python_sql_engine.md (hunk 15 above)
Comment (tobymao) [reply posts/python_sql_engine.md:77]: i plan to leave it in the mark down

File: posts/python_sql_engine.md [hunk 17]
```diff
@@ -0,0 +1,197 @@
...
+
+## Why?
+I started working on SQLGlot because of my work on the [experimentation and metrics platform](https://netflixtechblog.com/reimagining-experimentation-analysis-at-netflix-71356393af21) at Netflix, where I built tools that allowed data scientists to define and compute SQL based metrics. Netflix relied on multiple engines to query data (Spark, Presto, and Druid) so my team built the metrics platform around [PyPika](https://github.com/kayak/pypika), a Python SQL query builder so that definitions could be reused across multiple engines. However, it became quickly apparent that writing python code to programatically generate SQL was challenging for data scientists, especially those with academic backgrounds since they were mostly familiar with R and SQL. But at the time, the only Python SQL parser was [sqlparse]([https://github.com/andialbrecht/sqlparse) which is not actually a parser, but a tokenizer, so having users write raw SQL into the platform wasn't really an option. Some time later, I randomly stumbled across [Crafting Interpreters](https://craftinginterpreters.com/) and realized that I could use it as a guide towards creating my on SQL parser / transpiler.
```
Comment (izeigerman) [inline posts/python_sql_engine.md:17]: > creating my on SQL parser / transpiler.

typo - "my own".

File: posts/python_sql_engine.md [hunk 18]
```diff
@@ -0,0 +1,197 @@
...
+
+Why did I do this? Isn't a Python SQL engine going to be extremely slow?
+
+The main reason why I ended up building a SQL engine was ... just for **entertainment**. It's been fun learning about all the things required to actually run a SQL query and seeing it actually work is extremely rewarding. Before SQLGlot, I had zero experience with lexers, parsers, or compilers.
+
+In terms of practical use cases, I planned to use the Python SQL engine for unit testing SQL pipelines. Big data pipelines are tough to test because many of the engines are not open source and cannot be run locally. With SQLGlot, you can take a SQL query targeting a warehouse like [Snowflake](https://www.snowflake.com/en/), and seamlessly run it in CI on mock Python data. It's easy to mock data and create arbitrary [UDFs](https://en.wikipedia.org/wiki/User-defined_function) because everything is just Python.  Although the implementation is slow and unsuitable for large amounts of data (> 1 millon rows), there's very little overhead / startup and can run queries on test data in a couple of milliseconds.
```
Comment (izeigerman) [inline posts/python_sql_engine.md:23]: > and can run queries on test data in a couple of milliseconds.

and "it" can run? Otherwise the sentence feels a little off.

File: posts/python_sql_engine.md [hunk 19]
```diff
@@ -0,0 +1,197 @@
...
+
+Running the SQLGlot tokenizer shows the output.
+
+![Tokenizer Output](python_sql_engine_images/tokenizer.png)
+
+Each keyword has been converted to a SQLGlot Token object. Each token has some metadata associated with it like line / column information for error messages. Comments are also a part of the token so that comments can be preserved.
+
+## Parsing
+
+Once a SQL statement is tokenized, we don't need to worry about white space and other formatting, so it's easier to work with. We can now convert the list of tokens into an AST. The SQLGlot [parser](https://github.com/tobymao/sqlglot/blob/main/sqlglot/parser.py) is a hand written [recursive descent](https://en.wikipedia.org/wiki/Recursive_descent_parser) parser.
```
Comment (izeigerman) [inline posts/python_sql_engine.md:57]: > to worry about white space

white space(s) - plural? 

> the list of tokens into an AST

worth expanding AST in parenthesis (Abstract Syntax Tree) the first time you refer to it. Perhaps add a link to Wikipedia.

File: posts/python_sql_engine.md [hunk 20]
```diff
@@ -0,0 +1,197 @@
...
+
+Similar to the tokenizer, it consumes the tokens sequentially but it uses a recursive algorithm instead. The tokens are converted into a single AST node which presents the SQL query. The SQLGlot parser was designed to support various dialects so it contains many options for overriding parsing functionality.
+
+![Parser Output](python_sql_engine_images/parser.png)
+
+The AST is a generic representation of a given SQL query. Each dialect can override or implement its own generator that can convert an AST object into syntatically correct SQL.
```
Comment (izeigerman) [inline posts/python_sql_engine.md:63]: > into syntatically correct SQL.

"into a syntatically correct SQL expression."?

File: posts/python_sql_engine.md (hunk 19 above)
Comment (tobymao) [reply posts/python_sql_engine.md:57]: it's done earlier
